### PR TITLE
Pyreverse: Use dashed lines for type-checking imports

### LINKED ISSUE #8112: Idea: Pyreverse should distinguish between functional imports and type checking imports
### Current problem

Pyreverse generates a graph that depicts the import dependencies between modules. But it doesn't distinguish between imports that are used for runtime functionality and imports that are used only for type checking (that is, imports behind the `TYPE_CHECKING` flag).

### Desired solution

It would be neat if different kinds of import dependencies were distinguished on the graph. For example, regular imports could be notated with a straight line (as is currently the case), while type imports could be notated with a squiggly line. Then it would be possible to tell at a glance what sort of relationship different modules really have.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by nickdrozd**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description

This PR modifies the Pyreverse package diagram generator so that if module A *only* imports from module B under the `TYPE_CHECKING` flag, an empty arrow is used.

Closes #8112

Here is an example of the output:

packages

In this diagram, the `instrs` module is only imported for type checking, the `tape` module is sometimes imported for type checking and sometimes imported for real, and the rest of the modules are all imported for real.

There is an interesting interaction with the (still unreleased) [`--no-standalone` option](https://github.com/pylint-dev/pylint/pull/8520). Namely, a module only imported for type checking appears as a ghost module:

packages

I hadn't considered this in my implementation, but I think it actually works out pretty well.

## TODO

- [x] Documentation
- [x] Testing
- [x] Add output for writers other than dot.

### CODE REVIEW

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Those "empty arrows" have a meaning in UML it means generalization, i.e. "inherit from". The proper UML convention needs to be used, possibly dotted line with the same arrow for "dependency" see (14) here : https://stackoverflow.com/a/2293763/2519059.

**Comment by nickdrozd** [comment]
Using dashed lines instead for type-checking-only dependencies:

packages

**Comment by nickdrozd** [comment]
@DudeNr33 Something is off with the tests, but I can't what it is. There's some kind of interference with `clientmodule_test`, but I don't know where it's coming from.

File: pylint/pyreverse/diagrams.py [hunk 1]
```diff
@@ -304,3 +310,10 @@ def extract_relationships(self) -> None:
                 except KeyError:
                     continue
                 self.add_relationship(package_obj, dep, "depends")
+
+            for dep_name in package_obj.node.type_depends:
+                try:
+                    dep = self.get_module(dep_name, package_obj.node)
+                except KeyError:
+                    continue
```
Comment (nickdrozd) [inline pylint/pyreverse/diagrams.py:318]: These two lines are uncovered, and I don't know how to cover them. Any ideas? They could also be cut, or just ignore the coverage.

**Review by Pierre-Sassoulas — COMMENTED** [review]
I think this look dope, great change (especially in conjunction with the other option) ! Although I would appreciate a second / third opinion regarding the UML symbolic before we commit to it.

**Comment by nickdrozd** [comment]
Pyreverse package diagram:

packages

File: pylint/pyreverse/printer.py [hunk 2]
```diff
@@ -25,6 +25,7 @@ class EdgeType(Enum):
     ASSOCIATION = "association"
     AGGREGATION = "aggregation"
     USES = "uses"
+    TYPE_USES = "type_uses"
```
Comment (DudeNr33) [inline pylint/pyreverse/printer.py:28]:
```suggestion
    TYPE_DEPENDENCY = "type_dependency"
```
I think this gets the point a little bit better across, and is more consistent with the new `type_depends` attribute.

File: tests/pyreverse/conftest.py [hunk 3]
```diff
@@ -36,6 +36,13 @@ def no_standalone_dot_config() -> PyreverseConfig:
     )
 
 
+@pytest.fixture()
+def type_check_imports_dot_config() -> PyreverseConfig:
+    return PyreverseConfig(
+        output_format="dot",
+    )
```
Comment (DudeNr33) [inline tests/pyreverse/conftest.py:43]: This is the same as the `default_config`, so we might as well use that one!

**Review by DudeNr33 — CHANGES_REQUESTED** [review]
Thank you Nick for the new feature. I like it.

I left some remarks.

I also think we should try to introduce real functional package diagram tests rather sooner than later after this PR. The functional tests for class diagram work more or less the same as the functional tests for messages - that is, you only need to create the input, output, and optionally config file and they will be picked up automatically in the test run.
This is not yet the case for the files below `functional/package_diagram`, but the naming convention could make you believe it is.

**Review by Pierre-Sassoulas — APPROVED** [review]

**Review by DudeNr33 — APPROVED** [review]

**Comment by DudeNr33** [comment]
codecov seems to be stuck. @Pierre-Sassoulas what's the best fix for this? Close/reopen the PR?

**Comment by Pierre-Sassoulas** [comment]
Yeah, let's try !

**Comment by Pierre-Sassoulas** [comment]
There's some trickery with astroid's version but I'm not sure why readthedoc alone fail. Maybe bad caching. I'm half inclined to "just merge it ™️".

**Comment by Pierre-Sassoulas** [comment]
Waiting for https://github.com/pylint-dev/pylint/pull/7767 and a rebase might be best.

**Comment by jacobtylerwalls** [comment]
> There's some trickery with astroid's version but I'm not sure why readthedoc alone fail.

Yeah, I didn't bother to pin the astroid commit in the docs/requirements.txt, so it was pulling from astroid a7-dev0. Should be good now 👍

File: tests/pyreverse/functional/package_diagrams/type_check_imports/packages.mmd [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+  }
+  class mod_b {
+  }
+  class mod_c {
+  }
+  class mod_d {
+  }
+  mod_b --> mod_a
+  mod_d --> mod_a
+  mod_c -.-> mod_a
```
Comment (agusdmb) [inline tests/pyreverse/functional/package_diagrams/type_check_imports/packages.mmd:14]: This is not valid mermaid.

it should be:
`mod_c ..> mod_a`
as documented here:
https://mermaid.js.org/syntax/classDiagram.html#defining-relationship

github renders mermaid files, and if we try to open this one we get an error, see:
https://github.com/pylint-dev/pylint/blob/ffe1e714a3cfe83b5a627ade6d723489f2e10621/tests/pyreverse/functional/package_diagrams/type_check_imports/packages.mmd

File: pylint/pyreverse/mermaidjs_printer.py [hunk 5]
```diff
@@ -24,6 +24,7 @@ class MermaidJSPrinter(Printer):
         EdgeType.ASSOCIATION: "--*",
         EdgeType.AGGREGATION: "--o",
         EdgeType.USES: "-->",
+        EdgeType.TYPE_DEPENDENCY: "-.->",
```
Comment (agusdmb) [inline pylint/pyreverse/mermaidjs_printer.py:27]: this is the actual wrong line

Comment (jacobtylerwalls) [reply pylint/pyreverse/mermaidjs_printer.py:27]: Thanks! Would you be willing to provide a PR?

Comment (agusdmb) [reply pylint/pyreverse/mermaidjs_printer.py:27]: yep... just did!
#9031

**Pull request merged**